### DISCUSSION TITLE
Create objects with Graphql Datahub the userOwner property is set to user "system" (id = 0)

### DISCUSSION THREAD

**Discussion by Jimerson21** [pimcore/pimcore#8628] [Q&A / Support]
I am doing a login an existing “backend user” using a frontend view. This is my configuration: 

security.yml: 

    security:
        providers:
            pimcore_admin:
                id: Pimcore\Bundle\AdminBundle\Security\User\UserProvider
    
        firewalls:
    
            iqimagen:
                anonymous: ~
                pattern: ^/es_MX/ffpromos/
                provider: pimcore_admin
                guard:
                    entry_point: AppBundle\Security\Guard\AdminAuthenticator
                    authenticators:
                        - AppBundle\Security\Guard\AdminAuthenticator
                form_login:
                    login_path: user_login
                    check_path: user_login
                    use_forward: false
                logout:
                    path: user_logout
                    target: user_login
                    invalidate_session: false

AdminAuthenticator.php

      <?php 
      
      namespace AppBundle\Security\Guard;
      
      use Pimcore\Bundle\AdminBundle\Security\Authentication\Token\TwoFactorRequiredToken;
      use Pimcore\Bundle\AdminBundle\Security\User\User;
      use Pimcore\Tool\Authentication;
      use Pimcore\Tool\Session;
      use Symfony\Component\HttpFoundation\Cookie;
      use Symfony\Component\HttpFoundation\RedirectResponse;
      use Symfony\Component\HttpFoundation\Request;
      use Symfony\Component\HttpFoundation\Session\Attribute\AttributeBagInterface;
      use Symfony\Component\Routing\RouterInterface;
      use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
      use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
      use Symfony\Component\Security\Core\Exception\AuthenticationException;
      use Symfony\Component\Security\Core\Exception\InvalidCsrfTokenException;
      use Symfony\Component\Security\Core\User\UserInterface;
      use Symfony\Component\Security\Core\User\UserProviderInterface;
      use Symfony\Component\Security\Csrf\CsrfToken;
      use Symfony\Component\Security\Csrf\CsrfTokenManagerInterface;
      use Symfony\Component\Security\Guard\AbstractGuardAuthenticator;
      
      class AdminAuthenticator extends AbstractGuardAuthenticator
      {
          private $router;
          private $csrfTokenManager;
          private $tokenStorage;
          
          public function __construct(RouterInterface $router, CsrfTokenManagerInterface $csrfTokenManager, TokenStorageInterface $tokenStorage)
          {
              $this->router = $router;
              $this->csrfTokenManager = $csrfTokenManager;
              $this->tokenStorage = $tokenStorage;
          }
        
          public function supports(Request $request)
          {
              return $request->attributes->get('_route') === 'user_login' && $request->isMethod('POST');
          }
      
          public function getCredentials(Request $request)
          {
              $credentials = [];
              
              return $credentials = [
                  'username' => $request->request->get('_username'),
                  'password' => $request->request->get('_password'),
                  'csrf_token' => $request->request->get('_csrf_token')
              ];
          }
      
          public function getUser($credentials, UserProviderInterface $userProvider)
          {
              $user = null;
      
              if (!is_array($credentials)) {
                  throw new AuthenticationException('Invalid credentials');
              }
      
              $token = new CsrfToken('authenticate', $credentials['csrf_token']);
      
              if (!$this->csrfTokenManager->isTokenValid($token)) {
                  throw new InvalidCsrfTokenException();
              }
              
              if (isset($credentials['password'])) { 
      
                  $pimcoreUser = Authentication::authenticatePlaintext($credentials['username'], $credentials['password']);
      
                  if ($pimcoreUser) {
                   
                      $user = new User($pimcoreUser);
                      
                  }else{
      
                      return null;
                  }
              } 
      
              if ($user && Authentication::isValidUser($user->getUser())) {
                  
                  $pimcoreUser = $user->getUser();
      
                  Session::useSession(function (AttributeBagInterface $adminSession) use ($pimcoreUser, $token) {
                      Session::regenerateId();
                      $adminSession->set('user', $pimcoreUser);
                      $adminSession->set('csrfToken', $token->getValue());
                  });
              }
              return $user;
          }
      
          public function checkCredentials($credentials, UserInterface $user)
          {
              if ($user instanceof User) {
                  return true;
              }
      
              return false;
          }
      
          public function createAuthenticatedToken(UserInterface $user, $providerKey)
          {
              if ($this->twoFactorRequired) {
                  return new TwoFactorRequiredToken(
                      $user,
                      $providerKey,
                      $user->getRoles()
                  );
              } else {
                  return parent::createAuthenticatedToken($user, $providerKey);
              }
          }
      
          public function onAuthenticationFailure(Request $request, AuthenticationException $exception)
          {
              $url = $this->router->generate('user_logout', []);
      
              return new RedirectResponse($url);
          }
      
          public function onAuthenticationSuccess(Request $request, TokenInterface $token, $providerKey)
          {
              $user = $token->getUser();
              
              $this->tokenStorage->setToken($token);
              
              $url = $this->router->generate('dashboard', []);
                 
              if ($url) {
                  
                  $response = new RedirectResponse($url);
                  $response->headers->setCookie(new Cookie('pimcore_admin_sid', true, 0, '/', null, false, true));
      
                  return $response;
              }
      
              return null;
          }
      
          public function start(Request $request, AuthenticationException $authException = null)
          {
      
          }
      
          public function supportsRememberMe()
          {
              return false;
          }
          
      }

inside the frontend "user" is available

     $this->user = \Pimcore\Tool\Session::getReadonly()->get('user');

The problem is when I try to create an object using the graphql api. The userOwner property is always set to id 0, regardless of the logged in user. I have tried to replicate the AdminAuthenticator to get an authentication process similar to the Admin, but it did not work.

And this is because $currentUser = \PimcoreToolAdmin::getCurrentUser(); always returns null, because TokenStorageUserResolver is not storing the user's token or at least that's what I think.

Thank you.

**Comment by AlexR1712** [comment]
Same issue here

**Comment by gjorgic** [comment]
What is userOwner? is that property on your object? Did you cover graphQl endpoint with any firewall?

**Reply by Jimerson21** [reply]
userOwner is a system property built into pimcore objects, which is set automatically. 

And no, the API is not covered by any firewall, I just implemented a symfony authentication system as seen in my security.yml and adminAuthenticautor.php. That allows me to login with pimcore backend users but from my AppBundle. 

The problem comes given when creating the objects with the Graphql datahub API, the Pimcore\Tool\Admin :: getCurrentUser(); always returns null when the userOwner property is set. 

But when I create the objects with the PHP Api, it sets the property correctly, with the user logged in.

**Reply by gjorgic** [reply]
Correct me if i'm wrong, but pimcore can not know which user making request if you do not require user by firewall and  tell them to check which is;
Also that require you to send credentials somehow (_session, token... depence on firewall configuration_) while you requesting graphQl endpoint

**Reply by weisswurstkanone** [reply]
Maybe make use of this pre execute event and implement the logic there ?
https://github.com/pimcore/data-hub/blob/bbf0c3faa33e2bbc94e11f2797ef4eff5f6a0e76/src/Controller/WebserviceController.php#L160
